src/opm/input/eclipse/Schedule/KeywordHandlers.cpp

```diff
@@ -766,19 +766,69 @@ File {} line {}.)", wname, location.keyword, location.filename, location.lineno)
     }
 
     void Schedule::handleGRUPNET(HandlerContext& handlerContext) {
+        auto network = this->snapshots.back().network.get();
+        std::vector<Network::Node> nodes;
         for (const auto& record : handlerContext.keyword) {
-            const auto& groupName = record.getItem("NAME").getTrimmedString(0);
-
-            if (!this->snapshots.back().groups.has(groupName))
-                addGroup(groupName , handlerContext.currentStep);
-
-            int table = record.getItem("VFP_TABLE").get< int >(0);
+            const std::string& groupNamePattern = record.getItem<ParserKeywords::GRUPNET::NAME>().getTrimmedString(0);
+            const auto group_names = this->groupNames(groupNamePattern);
+            if (group_names.empty())
+                this->invalidNamePattern(groupNamePattern, handlerContext);
+            const auto& pressure_item = record.getItem<ParserKeywords::GRUPNET::TERMINAL_PRESSURE>();
+            const int vfp_table = record.getItem<ParserKeywords::GRUPNET::VFP_TABLE>().get<int>(0);
 
-            auto new_group = this->snapshots.back().groups.get( groupName );
-            if (new_group.updateNetVFPTable(table))
-                this->snapshots.back().groups.update( std::move(new_group) );
+            for (const auto& group_name : group_names) {
+                const auto& group = this->snapshots.back().groups.get(group_name);
+                const auto& parent_name = group.parent();
+                if (!parent_name.empty())
```

Comment:
The checking of the whether the node is a fixed-pressure node is at the later stage, it means that the code in the `if` condition will be gone through even if later we find the node is a fixed-pressure node. Although `throw` will let us know. 

I do see the logic regarding whether the node is a fixed pressure node and whether `vfp==0` is a little tangled, but for the ease of the code, I suggest the structure will be as follows, 

```c++
if (is_terminal_node) {
        throw if vfp > 0 (we should not have a valid vfp table for fixed pressure node);
        update node with pressure;
        remove the branch to the upstream node
} else { // not a terminal node
        if vfp <=0, remove the branch
        if vfp > 0 add the branch to the upstream node if there is valid parent group;
}

update the node with the information in this record if necessary;
update(network);
```

Reply:
I restructured the logic as you suggest which indeed makes the code more clear.